File: riveredge-backend/src/apps/kuaizhizao/schemas/sales_opportunity.py
"""
销售商机 Schema
"""

from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import List, Optional

from pydantic import Field

from core.schemas.base import BaseSchema


TERMINAL_STAGE_CODES = frozenset({"WON", "LOST"})


class SalesOpportunityBase(BaseSchema):
    """销售商机基础"""

    customer_id: int = Field(..., description="客户ID")
    customer_name: str = Field(..., max_length=200, description="客户名称")
    title: str = Field(..., max_length=200, description="商机名称")
    stage_code: str = Field(..., max_length=50, description="漏斗阶段字典值")
    status: str = Field("open", max_length=20, description="open / won / lost")
    expected_amount: Optional[Decimal] = Field(None, description="预计金额")
    expected_close_date: Optional[date] = Field(None, description="预计成交日期")
    owner_id: Optional[int] = Field(None, description="负责人ID")
    quotation_id: Optional[int] = Field(None, description="关联报价单ID")
    quotation_code: Optional[str] = Field(None, max_length=50, description="关联报价单编码")
    sales_order_id: Optional[int] = Field(None, description="关联销售订单ID")
    sales_order_code: Optional[str] = Field(None, max_length=50, description="关联销售订单编码")
    last_follow_up_at: Optional[datetime] = Field(None, description="最近跟进时间")
    next_follow_up_at: Optional[datetime] = Field(None, description="计划下次跟进时间")


class SalesOpportunityCreate(BaseSchema):
    """创建销售商机"""

    customer_id: int = Field(..., description="客户ID")
    title: str = Field(..., max_length=200, description="商机名称")
    stage_code: str = Field("INITIAL", max_length=50, description="初始阶段")
    expected_amount: Optional[Decimal] = Field(None, description="预计金额")
    expected_close_date: Optional[date] = Field(None, description="预计成交日期")
    quotation_id: Optional[int] = Field(None, description="关联报价单ID")
    sales_order_id: Optional[int] = Field(None, description="关联销售订单ID")


class SalesOpportunityUpdate(BaseSchema):
    """更新销售商机"""

    title: Optional[str] = Field(None, max_length=200, description="商机名称")
    stage_code: Optional[str] = Field(None, max_length=50, description="漏斗阶段")
    expected_amount: Optional[Decimal] = Field(None, description="预计金额")
    expected_close_date: Optional[date] = Field(None, description="预计成交日期")
    quotation_id: Optional[int] = Field(None, description="关联报价单ID")
    sales_order_id: Optional[int] = Field(None, description="关联销售订单ID")
    next_follow_up_at: Optional[datetime] = Field(None, description="计划下次跟进时间")


class SalesOpportunityEnsure(BaseSchema):
    """确保存在 open 商机（单据关联或客户默认跟进载体）"""

    customer_id: int = Field(..., description="客户ID")
    quotation_id: Optional[int] = Field(None, description="关联报价单ID")
    sales_order_id: Optional[int] = Field(None, description="关联销售订单ID")
    title: Optional[str] = Field(None, max_length=200, description="新建商机标题（可选）")


class SalesOpportunityResponse(SalesOpportunityBase):
    """销售商机响应"""

    id: int = Field(..., description="商机ID")
    uuid: str = Field(..., max_length=36, description="业务UUID")
    tenant_id: int = Field(..., description="租户ID")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")
    created_by: Optional[int] = Field(None, description="创建人ID")
    updated_by: Optional[int] = Field(None, description="更新人ID")


class SalesOpportunityListEnvelope(BaseSchema):
    """商机列表"""

    items: List[SalesOpportunityResponse] = Field(default_factory=list, description="当前页数据")
    total: int = Field(0, description="总条数")

File: riveredge-frontend/src/apps/kuaizhizao/services/sales-opportunity.ts
/**
 * 销售商机 API
 */

import { apiRequest } from '../../../services/api';

export interface SalesOpportunity {
  id: number;
  uuid?: string;
  tenant_id?: number;
  customer_id: number;
  customer_name: string;
  title: string;
  stage_code: string;
  status: 'open' | 'won' | 'lost' | string;
  expected_amount?: string | number | null;
  expected_close_date?: string | null;
  owner_id?: number | null;
  quotation_id?: number | null;
  quotation_code?: string | null;
  sales_order_id?: number | null;
  sales_order_code?: string | null;
  last_follow_up_at?: string | null;
  next_follow_up_at?: string | null;
  created_at?: string;
  updated_at?: string;
}

export interface SalesOpportunityListResult {
  items: SalesOpportunity[];
  total: number;
}

export const salesOpportunityApi = {
  list: async (params?: {
    skip?: number;
    limit?: number;
    customer_id?: number;
    status?: string;
  }): Promise<SalesOpportunityListResult> =>
    apiRequest('/apps/kuaizhizao/sales-opportunities', { method: 'GET', params }),

  get: async (id: number): Promise<SalesOpportunity> =>
    apiRequest(`/apps/kuaizhizao/sales-opportunities/${id}`, { method: 'GET' }),

  create: async (data: {
    customer_id: number;
    title: string;
    stage_code?: string;
    expected_amount?: number | null;
    expected_close_date?: string | null;
    quotation_id?: number | null;
    sales_order_id?: number | null;
  }): Promise<SalesOpportunity> =>
    apiRequest('/apps/kuaizhizao/sales-opportunities', { method: 'POST', data }),

  ensure: async (data: {
    customer_id: number;
    quotation_id?: number | null;
    sales_order_id?: number | null;
    title?: string | null;
  }): Promise<SalesOpportunity> =>
    apiRequest('/apps/kuaizhizao/sales-opportunities/ensure', { method: 'POST', data }),

  update: async (
    id: number,
    data: Partial<{
      title: string;
      stage_code: string;
      expected_amount: number | null;
      expected_close_date: string | null;
      quotation_id: number | null;
      sales_order_id: number | null;
      next_follow_up_at: string | null;
    }>
  ): Promise<SalesOpportunity> =>
    apiRequest(`/apps/kuaizhizao/sales-opportunities/${id}`, { method: 'PATCH', data }),
};
